File: app/src/main/java/net/xaethos/trackernotifier/adapters/ResourceAdapter.java
package net.xaethos.trackernotifier.adapters;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import net.xaethos.trackernotifier.models.Resource;

public class ResourceAdapter<R extends Resource, DS extends ResourceDataSource<R>>
        extends RecyclerView.Adapter<CommentViewHolder> implements DataSource.Observer {

    private LayoutInflater mLayoutInflater;
    private final DS mDataSource;

    ResourceAdapter(DS dataSource) {
        super();
        mDataSource = dataSource;
    }

    public DS getDataSource() {
        return mDataSource;
    }

    @Override
    public void onAttachedToRecyclerView(RecyclerView recyclerView) {
        mLayoutInflater = LayoutInflater.from(recyclerView.getContext());
        super.onAttachedToRecyclerView(recyclerView);
    }

    @Override
    public void onDetachedFromRecyclerView(RecyclerView recyclerView) {
        super.onDetachedFromRecyclerView(recyclerView);
        mLayoutInflater = null;
    }

    @Override
    public CommentViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new CommentViewHolder(mLayoutInflater.inflate(viewType, parent, false));
    }

    @Override
    public void onBindViewHolder(CommentViewHolder holder, int position) {
        holder.bind(mDataSource.getResource(position));
    }

    @Override
    public long getItemId(int position) {
        return mDataSource.getItemId(position);
    }

    @Override
    public int getItemCount() {
        return mDataSource.getItemCount();
    }

    @Override
    public int getItemViewType(int position) {
        return mDataSource.getItemViewType(position);
    }

}

File: app/src/main/java/net/xaethos/trackernotifier/adapters/ResourceAdapter.kt
package net.xaethos.trackernotifier.adapters

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import net.xaethos.trackernotifier.models.Resource
import java.util.*

class ResourceAdapter<R : Resource, VH : ResourceViewHolder<R>>(
        val itemViewType: Int = 0,
        val viewHolderFactory: (LayoutInflater, Int, ViewGroup) -> VH) : RecyclerView.Adapter<VH>() {

    private val items = ArrayList<R>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
            viewHolderFactory(LayoutInflater.from(parent.context), viewType, parent);

    override fun onBindViewHolder(holder: VH, position: Int) = holder.bind(get(position))

    override fun getItemId(position: Int) = get(position).id

    override fun getItemViewType(position: Int) = itemViewType

    override fun getItemCount() = items.size

    operator fun get(position: Int) = items[position]

    fun setResources(resources: Collection<R>?) {
        items.clear()
        if (resources != null) items.addAll(resources)
        notifyDataSetChanged()
    }

}
